### PR TITLE
CPI without Account Refs

### PR DESCRIPTION
**Pull request by Lichtso**
#### Problem
#15410 introduced [this](https://github.com/solana-labs/solana/blob/7a785e067ac109bb065ea1170af4bd9f896adfc2/runtime/src/message_processor.rs#L178) `unsafe` expression.

#### Summary of Changes
* Removes search for accounts and `unsafe` lifetime transmute in `InvokeContext::push()`.
* Replaces account parameters of all CPI methods by indices, as they are easier to verify and compatible with ABIv2 (see #19191).

Fixes #

### CODE REVIEW

File: runtime/src/message_processor.rs [hunk 1]
```diff
@@ -214,24 +194,63 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
...
-        )
+        let accounts = &self.accounts;
+        let pre_accounts = &mut self.pre_accounts;
+        let timings = &mut self.timings;
+
+        // Verify the per-account instruction results
+        let (mut pre_sum, mut post_sum) = (0_u128, 0_u128);
+        let mut work = |_unique_index: usize, index_in_instruction: usize| {
+            if index_in_instruction < write_privileges.len()
+                && index_in_instruction < accounts.len()
```
Comment (jstarry) [inline runtime/src/message_processor.rs:215]:
```suggestion
                && index_in_instruction < account_indices.len()
```
I think this should be updated

**Review by jstarry — COMMENTED** [review]
Nice cleanup, noticed something while reading your change

File: runtime/src/message_processor.rs (hunk 1 above)
Comment (Lichtso) [reply runtime/src/message_processor.rs]: Good find!

**Review by jstarry — APPROVED** [review]
Looks great, it also looks like this change makes https://github.com/solana-labs/solana/issues/9711 trivial to implement

**Pull request merged**